## TALLER_1_ANALISIS EXPLORATORIO DE DATOS

### Estudiante = Claudia Angel Rodriguez 

### Importar librerias 

In [302]:
import pandas as pd
import numpy as np
import matplotlib as plt
from matplotlib import pyplot
import matplotlib.pyplot as plt
from matplotlib import rcParams
import seaborn as sns
from datetime import date
from sklearn.preprocessing import OneHotEncoder
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor

### Contexto del análisis

La empresa cuya misión es la de importar y distribuir al por mayor lujos, accesorios y audio para el sector automotor a nivel Nacional, cuenta con una dataset donde tiene registrado el total de clientes que ha obtenido desde su creación hasta lo que se lleva del año 2023. La Gerente quiere conocer cual a sido el comportamiento de crecimiento del número de clientes desde que esta constituida hasta lo que se lleva del año 2023 y el desempeño de los vendedores en su captación. Esta dataset es una data estructurada donde cada fila representa un cliente y cada columna una variable característica del cliente.

### Pregunta SMART

Incremento de clientes en la empresa desde el año 2020 hasta febrero de 2023 y el desempeño de los vendedores en su captación.

### Cargar la dataset con sus encabezados y mostrar la primera parte de la misma

In [303]:
data = pd.read_excel("CLIENTES.xls")
# Asigno los encabezado
headers = ["Cdigo", "Razn social", "Sucursal", "Vendedor", "Razn social vendedor", "Fecha ingreso", "Antiguedad", "Calificacion", "Condicion de pago", "Desc. condicion de pago", "Cupo de crdito", "Tipo cliente", "Desc. tipo cliente", "Grupo dscto.", "Bloqueado", "Bloqueado cupo", "Bloqueado mora"
]
# Visualizar las 5 primeras filas de la data frame
data.head(5)

Cdigo                 Razn social  Sucursal  Vendedor  \
0   38643278    PARDO OSORIO LUZ ADRIANA         1       1.0   
1  901242233   POLARIZADOS SOLUTIONS SAS         3      11.0   
2  901242233   POLARIZADOS SOLUTIONS SAS         1       2.0   
3  901242233   POLARIZADOS SOLUTIONS SAS         2       1.0   
4  811039882  COMERCIALIZADORA JAGIR SAS         1      18.0   

           Razn social vendedor Fecha ingreso     Antiguedad Calificacion  \
0       DAVALOS PARDO ALEJANDRO    2020-10-01     2a  5m  5d            A   
1    JIMENEZ OVIEDO JOSE JULIAN    2021-07-14     1a  8m -8d            A   
2      ENCISO PARDO LUIS CARLOS    2020-10-01     2a  5m  5d            A   
3       DAVALOS PARDO ALEJANDRO    2021-02-02     2a  1m  4d            A   
4  GONZALEZ GOMEZ ANDRES FELIPE    2020-10-01     2a  5m  5d            A   

  Condicion de pago Desc. condicion de pago  Cupo de crdito  Tipo cliente  \
0               30D         CREDITO 30 DIAS               0             1   
1               30D         CREDITO 30 DIAS        15000000             1   
2               90D         CREDITO 90 DIAS        40500000             1   
3               30D         CREDITO 30 DIAS        25000000             1   
4               30D         CREDITO 30 DIAS        30000000             1   

    Desc. tipo cliente  Grupo dscto. Bloqueado Bloqueado cupo Bloqueado mora  
0  CLIENTES NACIONALES           NaN        No             Si             Si  
1  CLIENTES NACIONALES           1.0        No             Si             Si  
2  CLIENTES NACIONALES           1.0        No             No             No  
3  CLIENTES NACIONALES           NaN        No             Si             Si  
4  CLIENTES NACIONALES           1.0        No             Si             Si

### Descripción de los datos incluyendo para cada columna: Variable: nombre de la variable

#### Descripción de los campos

Cada fila contiene la información del cliente desde su ingreso hasta la definición del tipo de cliente, se describe a continuación:

**Cdigo**: id cliente

**Razn social**: Nombre del cliente como persona jurídica

**Sucursal**:Zona donde se encuentra el cliente

**Vendedor**: id vendedor

**Razn social vendedor**:Nombre del Vendedor que atiende al cliente

**Fecha ingreso**: Fecha en la que ingreso el cliente al sistema de la empresa

**Antiguedad**: cantidad de años que tiene el cliente en la empresa

**Calificacion**: calificación del cliente en función de la documentacion de ingreso

**Condicion de pago**: esta indica si es de contado o a crédito, y respecto a crédito el tipo de plazo de pago

**Desc. condicion de pago**: indica el descuento en función de la condición de pago

**Cupo de crdito**: monto de dinero que la empresa le aprueba como crédito al cliente

**Tipo cliente**: tipo de cliente de acuerdo a la calificación obtenida a partir de la documentación

**Desc. tipo cliente**:define el descuento de acuerdo al tipo de cliente

**Grupo dscto.**: define el grupo donde se encuentra el cliente de acuerdo al descuento de la condición de pago

**Bloqueado**: indica si el cliente ya no es parte de la empresa

**Bloqueado cupo**: indica bloqueo de cupo para el cliente

**Bloqueado mora**: indica bloqueo del crédito del cliente por mora

### Contexto (información útil para entender el espacio semántico). 

Para dar respuesta a la pregunta SMART, se tomaran del dataset las siguientes variables:  

 **Cdigo**: id cliente
 
 **Vendedor**: id vendedor
 
 **Fecha ingreso**: Fecha en la que ingreso el cliente al sistema de la empresa
 
 **Bloqueado**: indica si el cliente ya no es parte de la empresa

### Expectativa (Por qué es importante esta variable para mi problema)

Debido a que la pregunta SMART se enfoca en el incremento de los cliente en el tiempo de operación de la empresa y además que papel ha jugado cada vendedor en este indicador, a partir del id del cliente sabriamos la cantidad de clientes ingresados, la fecha de ingreso indicaria en que tiempo ingreso a la empresa, a partir de la variable vendedor como su nombre lo indica, se cruzaria con la cantidad de clientes que a captado para la empresa y finalmente a partir de la variable Bloqueado, se sabria que clientes han ingresado pero ya no hacen parte de los clientes actuales de la empresa.

### Importancia

Entender como ha sido el comportamiento en la consecución de los clientes durante la operación de la empresa en el mercado.

### Análisis general

#### Forma y tamaño del dataset

In [304]:
# forma de la Dataset: es decir cuenta con 17 columnas(Caracteristicas de los clientes) y 2221 filas(Clientes)
data.shape

(2221, 17)

In [305]:
# Tamaño de la Dataset: En total cuenta con 37757 datos.
data.size

37757

#### Tipos de variables

In [306]:
print(data.dtypes)

Cdigo                              object
Razn social                        object
Sucursal                            int64
Vendedor                          float64
Razn social vendedor               object
Fecha ingreso              datetime64[ns]
Antiguedad                         object
Calificacion                       object
Condicion de pago                  object
Desc. condicion de pago            object
Cupo de crdito                      int64
Tipo cliente                        int64
Desc. tipo cliente                 object
Grupo dscto.                      float64
Bloqueado                          object
Bloqueado cupo                     object
Bloqueado mora                     object
dtype: object


El total de variables del dataset son de tipo object, int, float y datatime. Y respecto a las variables que se usaran para dar respuesta a la pregunta SMART:

**Cdigo** (Cliente): object

**Vendedor**: float
  
 **Fecha ingreso**: datatime
 
 **Bloqueado**: object
 
Por otro lado tambien se logra identificar que variables como Sucursal, Tipo de cliente, y grupo de descuento aparecen como variables int, indicando que estas no estan bien clasificadas ya que en realidad son variables cualitativas nominales, ya que son variables categóricas y de un unico valor.

La estídistica de las variables cuantitativas de la dataset se presenta a continuación:

In [307]:
data.describe()

Sucursal     Vendedor  Cupo de crdito  Tipo cliente  Grupo dscto.
count  2221.000000  2220.000000    2.221000e+03   2221.000000    601.000000
mean      1.010356    16.334234    2.779611e+06      1.000450      1.001664
std       0.117714   299.889709    7.002404e+06      0.021219      0.040791
min       0.000000     1.000000    0.000000e+00      1.000000      1.000000
25%       1.000000     2.000000    0.000000e+00      1.000000      1.000000
50%       1.000000     5.000000    0.000000e+00      1.000000      1.000000
75%       1.000000    13.000000    3.000000e+06      1.000000      1.000000
max       3.000000  9999.000000    1.400000e+08      2.000000      2.000000

Ahora queremos conocer la estadística de todas las variables incluyendo las de tipo object, esto se describe a continuación:

In [356]:
data.describe(include ='all',datetime_is_numeric=True)

Cdigo                Razn social     Sucursal  Vendedor  \
count        2220.0                       2220  2220.000000    2220.0   
unique       2198.0                       2198          NaN      19.0   
top     901242233.0  POLARIZADOS SOLUTIONS SAS          NaN       1.0   
freq            3.0                          3          NaN     465.0   
mean            NaN                        NaN     1.010360       NaN   
min             NaN                        NaN     0.000000       NaN   
25%             NaN                        NaN     1.000000       NaN   
50%             NaN                        NaN     1.000000       NaN   
75%             NaN                        NaN     1.000000       NaN   
max             NaN                        NaN     3.000000       NaN   
std             NaN                        NaN     0.117741       NaN   

           Razn social vendedor                  Fecha ingreso     Antiguedad  \
count                      2220                           2220           2220   
unique                       19                            NaN            258   
top     DAVALOS PARDO ALEJANDRO                            NaN     2a  5m  5d   
freq                        465                            NaN           1524   
mean                        NaN  2021-02-09 13:55:27.567567360            NaN   
min                         NaN            2020-10-01 00:00:00            NaN   
25%                         NaN            2020-10-01 00:00:00            NaN   
50%                         NaN            2020-10-01 00:00:00            NaN   
75%                         NaN            2021-03-08 00:00:00            NaN   
max                         NaN            2023-03-03 00:00:00            NaN   
std                         NaN                            NaN            NaN   

       Calificacion Condicion de pago Desc. condicion de pago  Cupo de crdito  \
count          2220              2214                    2214    2.220000e+03   
unique            1                11                      10             NaN   
top               A               01D          CREDITO 01 DIA             NaN   
freq           2220              1037                    1037             NaN   
mean            NaN               NaN                     NaN    2.780863e+06   
min             NaN               NaN                     NaN    0.000000e+00   
25%             NaN               NaN                     NaN    0.000000e+00   
50%             NaN               NaN                     NaN    0.000000e+00   
75%             NaN               NaN                     NaN    3.000000e+06   
max             NaN               NaN                     NaN    1.400000e+08   
std             NaN               NaN                     NaN    7.003733e+06   

        Tipo cliente   Desc. tipo cliente  Grupo dscto. Bloqueado  \
count    2220.000000                 2220    601.000000      2220   
unique           NaN                    2           NaN         2   
top              NaN  CLIENTES NACIONALES           NaN        No   
freq             NaN                 2219           NaN      2212   
mean        1.000450                  NaN      1.001664       NaN   
min         1.000000                  NaN      1.000000       NaN   
25%         1.000000                  NaN      1.000000       NaN   
50%         1.000000                  NaN      1.000000       NaN   
75%         1.000000                  NaN      1.000000       NaN   
max         2.000000                  NaN      2.000000       NaN   
std         0.021224                  NaN      0.040791       NaN   

       Bloqueado cupo Bloqueado mora     Mes     Año  
count            2220           2220  2220.0  2220.0  
unique              2              2    12.0     4.0  
top                Si             Si    10.0  2020.0  
freq             1882           1899  1575.0  1529.0  
mean              NaN            NaN     NaN     NaN  
min               NaN           

### Remplazar valores perdidos

Ahora vamos a identificar si existen valores "?" que significa valores perdidos, y procederemos a convertirlos en "NaN"

In [309]:
# Reemplazar "?" por NaN, esta función no se ejecuta ya que no tengo valores perdidos
data.replace("?", np.nan, inplace = True)
data.head(5)

Cdigo                 Razn social  Sucursal  Vendedor  \
0   38643278    PARDO OSORIO LUZ ADRIANA         1       1.0   
1  901242233   POLARIZADOS SOLUTIONS SAS         3      11.0   
2  901242233   POLARIZADOS SOLUTIONS SAS         1       2.0   
3  901242233   POLARIZADOS SOLUTIONS SAS         2       1.0   
4  811039882  COMERCIALIZADORA JAGIR SAS         1      18.0   

           Razn social vendedor Fecha ingreso     Antiguedad Calificacion  \
0       DAVALOS PARDO ALEJANDRO    2020-10-01     2a  5m  5d            A   
1    JIMENEZ OVIEDO JOSE JULIAN    2021-07-14     1a  8m -8d            A   
2      ENCISO PARDO LUIS CARLOS    2020-10-01     2a  5m  5d            A   
3       DAVALOS PARDO ALEJANDRO    2021-02-02     2a  1m  4d            A   
4  GONZALEZ GOMEZ ANDRES FELIPE    2020-10-01     2a  5m  5d            A   

  Condicion de pago Desc. condicion de pago  Cupo de crdito  Tipo cliente  \
0               30D         CREDITO 30 DIAS               0             1   
1               30D         CREDITO 30 DIAS        15000000             1   
2               90D         CREDITO 90 DIAS        40500000             1   
3               30D         CREDITO 30 DIAS        25000000             1   
4               30D         CREDITO 30 DIAS        30000000             1   

    Desc. tipo cliente  Grupo dscto. Bloqueado Bloqueado cupo Bloqueado mora  
0  CLIENTES NACIONALES           NaN        No             Si             Si  
1  CLIENTES NACIONALES           1.0        No             Si             Si  
2  CLIENTES NACIONALES           1.0        No             No             No  
3  CLIENTES NACIONALES           NaN        No             Si             Si  
4  CLIENTES NACIONALES           1.0        No             Si             Si

#### Analizar los Valores únicos de las variables discretas

A continuación se muestra que no existen valores perdidos en la dataset (False:Valores no perdidos)

In [310]:
missing_data = data.isnull()
missing_data.head(5)

Cdigo  Razn social  Sucursal  Vendedor  Razn social vendedor  \
0  False        False     False     False                 False   
1  False        False     False     False                 False   
2  False        False     False     False                 False   
3  False        False     False     False                 False   
4  False        False     False     False                 False   

   Fecha ingreso  Antiguedad  Calificacion  Condicion de pago  \
0          False       False         False              False   
1          False       False         False              False   
2          False       False         False              False   
3          False       False         False              False   
4          False       False         False              False   

   Desc. condicion de pago  Cupo de crdito  Tipo cliente  Desc. tipo cliente  \
0                    False           False         False               False   
1                    False           False         False               False   
2                    False           False         False               False   
3                    False           False         False               False   
4                    False           False         False               False   

   Grupo dscto.  Bloqueado  Bloqueado cupo  Bloqueado mora  
0          True      False           False           False  
1         False      False           False           False  
2         False      False           False           False  
3          True      False           False           False  
4         False      False           False           False

Para contar posibles valores númericos aplicamos un For que nos cuente cuantos true aparecen, debido a que no se tienen valores perdidos, solo cuenta valores False, lo mencionado anteriormente se presenta a continuación:

In [311]:
for column in missing_data.columns.values.tolist():
    print(column) 
    print (missing_data[column].value_counts())
    print("") 

Cdigo
False    2221
Name: Cdigo, dtype: int64

Razn social
False    2221
Name: Razn social, dtype: int64

Sucursal
False    2221
Name: Sucursal, dtype: int64

Vendedor
False    2220
True        1
Name: Vendedor, dtype: int64

Razn social vendedor
False    2220
True        1
Name: Razn social vendedor, dtype: int64

Fecha ingreso
False    2221
Name: Fecha ingreso, dtype: int64

Antiguedad
False    2221
Name: Antiguedad, dtype: int64

Calificacion
False    2221
Name: Calificacion, dtype: int64

Condicion de pago
False    2214
True        7
Name: Condicion de pago, dtype: int64

Desc. condicion de pago
False    2214
True        7
Name: Desc. condicion de pago, dtype: int64

Cupo de crdito
False    2221
Name: Cupo de crdito, dtype: int64

Tipo cliente
False    2221
Name: Tipo cliente, dtype: int64

Desc. tipo cliente
False    2221
Name: Desc. tipo cliente, dtype: int64

Grupo dscto.
True     1620
False     601
Name: Grupo dscto., dtype: int64

Bloqueado
False    2221
Name: Bloqueado, dtype

Sin embargo la dataset si traiga consigo valores NaN, para este caso se quiere saber cuantos de estos valores se tiene por cada columna:

In [312]:
data.isna().sum()

Cdigo                         0
Razn social                   0
Sucursal                      0
Vendedor                      1
Razn social vendedor          1
Fecha ingreso                 0
Antiguedad                    0
Calificacion                  0
Condicion de pago             7
Desc. condicion de pago       7
Cupo de crdito                0
Tipo cliente                  0
Desc. tipo cliente            0
Grupo dscto.               1620
Bloqueado                     0
Bloqueado cupo                0
Bloqueado mora                0
dtype: int64

In [313]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2221 entries, 0 to 2220
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   Cdigo                    2221 non-null   object        
 1   Razn social              2221 non-null   object        
 2   Sucursal                 2221 non-null   int64         
 3   Vendedor                 2220 non-null   float64       
 4   Razn social vendedor     2220 non-null   object        
 5   Fecha ingreso            2221 non-null   datetime64[ns]
 6   Antiguedad               2221 non-null   object        
 7   Calificacion             2221 non-null   object        
 8   Condicion de pago        2214 non-null   object        
 9   Desc. condicion de pago  2214 non-null   object        
 10  Cupo de crdito           2221 non-null   int64         
 11  Tipo cliente             2221 non-null   int64         
 12  Desc. tipo cliente       2221 non-

Las columnas que tienen valores NaN son Vendedor, con un NaN, Razn social venderdor con un NaN, Condicion de pago con 7 NaN, Desc. condicion de pago con 7 NaN y Grupo dscto. con 1620 NaN. Sin embargo si los comparamos con los valores No NaN (non-null) que se presenta en la tabla superior, solo en la columna Grupo dscto se puede ver que la diferencia es bastante significativa contando solo con el 27% de sus datos, peso debido a que esta columna no es objeto para la pregunta SMART, no se le aplicara ningún tratamiento. Caso contrario con la columna vendedor donde se encontro tambien NaN procederemos a identificar que hacer con esta en la fase de Lidiar con valores perdidos.

### Lidiar con valores perdidos

En vista que la columna Vendedor tiene un NaN, siendo este tan solo el 0.045% del total de los datos de la columna, procederemos a eliminar la fila que contiene este valor:

In [314]:
# Eliminar la fila completa que tiene valor NaN en la columna "Vendedor" 
data.dropna(subset=["Vendedor"], axis=0, inplace=True)
# Restablecer el índice del dataframe porque se elimino una fila
data.reset_index(drop=True, inplace=True)
data.head(5)

Cdigo                 Razn social  Sucursal  Vendedor  \
0   38643278    PARDO OSORIO LUZ ADRIANA         1       1.0   
1  901242233   POLARIZADOS SOLUTIONS SAS         3      11.0   
2  901242233   POLARIZADOS SOLUTIONS SAS         1       2.0   
3  901242233   POLARIZADOS SOLUTIONS SAS         2       1.0   
4  811039882  COMERCIALIZADORA JAGIR SAS         1      18.0   

           Razn social vendedor Fecha ingreso     Antiguedad Calificacion  \
0       DAVALOS PARDO ALEJANDRO    2020-10-01     2a  5m  5d            A   
1    JIMENEZ OVIEDO JOSE JULIAN    2021-07-14     1a  8m -8d            A   
2      ENCISO PARDO LUIS CARLOS    2020-10-01     2a  5m  5d            A   
3       DAVALOS PARDO ALEJANDRO    2021-02-02     2a  1m  4d            A   
4  GONZALEZ GOMEZ ANDRES FELIPE    2020-10-01     2a  5m  5d            A   

  Condicion de pago Desc. condicion de pago  Cupo de crdito  Tipo cliente  \
0               30D         CREDITO 30 DIAS               0             1   
1               30D         CREDITO 30 DIAS        15000000             1   
2               90D         CREDITO 90 DIAS        40500000             1   
3               30D         CREDITO 30 DIAS        25000000             1   
4               30D         CREDITO 30 DIAS        30000000             1   

    Desc. tipo cliente  Grupo dscto. Bloqueado Bloqueado cupo Bloqueado mora  
0  CLIENTES NACIONALES           NaN        No             Si             Si  
1  CLIENTES NACIONALES           1.0        No             Si             Si  
2  CLIENTES NACIONALES           1.0        No             No             No  
3  CLIENTES NACIONALES           NaN        No             Si             Si  
4  CLIENTES NACIONALES           1.0        No             Si             Si

In [315]:
#Verifico que se haya eliminado la fila del dataset
data.shape

(2220, 17)

### Corregir formato de datos

Para el caso de las variables que se utilizaran para dar respuesta a la pregunta SMART 

**Cdigo**: object
 
 **Vendedor**: float
 
 **Fecha ingreso**: datatime
 
 **Bloqueado**: object

Como se quiere identificar la consecución de clientes por mes y año, se debe crear las columnas "Mes" y "Año" a partir de la columna "Fecha de ingreso".

In [316]:
data["Mes"] = data["Fecha ingreso"].dt.month
data["Año"] = data["Fecha ingreso"].dt.year
data.head(5)

Cdigo                 Razn social  Sucursal  Vendedor  \
0   38643278    PARDO OSORIO LUZ ADRIANA         1       1.0   
1  901242233   POLARIZADOS SOLUTIONS SAS         3      11.0   
2  901242233   POLARIZADOS SOLUTIONS SAS         1       2.0   
3  901242233   POLARIZADOS SOLUTIONS SAS         2       1.0   
4  811039882  COMERCIALIZADORA JAGIR SAS         1      18.0   

           Razn social vendedor Fecha ingreso     Antiguedad Calificacion  \
0       DAVALOS PARDO ALEJANDRO    2020-10-01     2a  5m  5d            A   
1    JIMENEZ OVIEDO JOSE JULIAN    2021-07-14     1a  8m -8d            A   
2      ENCISO PARDO LUIS CARLOS    2020-10-01     2a  5m  5d            A   
3       DAVALOS PARDO ALEJANDRO    2021-02-02     2a  1m  4d            A   
4  GONZALEZ GOMEZ ANDRES FELIPE    2020-10-01     2a  5m  5d            A   

  Condicion de pago Desc. condicion de pago  Cupo de crdito  Tipo cliente  \
0               30D         CREDITO 30 DIAS               0             1   
1               30D         CREDITO 30 DIAS        15000000             1   
2               90D         CREDITO 90 DIAS        40500000             1   
3               30D         CREDITO 30 DIAS        25000000             1   
4               30D         CREDITO 30 DIAS        30000000             1   

    Desc. tipo cliente  Grupo dscto. Bloqueado Bloqueado cupo Bloqueado mora  \
0  CLIENTES NACIONALES           NaN        No             Si             Si   
1  CLIENTES NACIONALES           1.0        No             Si             Si   
2  CLIENTES NACIONALES           1.0        No             No             No   
3  CLIENTES NACIONALES           NaN        No             Si             Si   
4  CLIENTES NACIONALES           1.0        No             Si             Si   

   Mes   Año  
0   10  2020  
1    7  2021  
2   10  2020  
3    2  2021  
4   10  2020

In [317]:
#Debido a que las variables Mes y Año estan en formato int, se procede  a corregir a formato object,
#De igual forma se corrige el formato de Vendedor de float a object
data.dtypes
#Cambiamos los formatos de los atributos  
data[["Mes"]] = data[["Mes"]].astype("object")
data[["Año"]] = data[["Año"]].astype("object")
data[["Vendedor"]] = data[["Vendedor"]].astype("object")
data.dtypes

Cdigo                              object
Razn social                        object
Sucursal                            int64
Vendedor                           object
Razn social vendedor               object
Fecha ingreso              datetime64[ns]
Antiguedad                         object
Calificacion                       object
Condicion de pago                  object
Desc. condicion de pago            object
Cupo de crdito                      int64
Tipo cliente                        int64
Desc. tipo cliente                 object
Grupo dscto.                      float64
Bloqueado                          object
Bloqueado cupo                     object
Bloqueado mora                     object
Mes                                object
Año                                object
dtype: object

#### Analizar  las variables discretas objeto de la pregunta SMART

#### Variable = Cdigo (cliente)

In [318]:
# Para dar respuesta a la pregunta SMART se requiere identificar el numero de clientes que ingresaron 
# a la empresa por año, para este caso se aplica la función de conteo de datos unicos y la función de eliminar 
#los datos duplicados:
data.groupby('Año')['Cdigo'].nunique()
Clientes_año = data.drop_duplicates(subset=['Cdigo']).groupby('Año')['Cdigo'].nunique()
Clientes_año

Año
2020    1520
2021     368
2022     267
2023      43
Name: Cdigo, dtype: int64

In [350]:
#Para determinar el número de clientes por mes y por año, aplicamos un filtro por año:
año=2023
año_conteo = data['Año'] == año
año_filtro = data[año_conteo]
#año_filtro.head()

In [351]:
# De igual forma se requiere identificar el numero de clientes que ingresaron por mes: para este caso lo aplicaremos 
# con el filtro por año
año_filtro.groupby('Mes')['Cdigo'].nunique()
Clientes_mes = año_filtro.drop_duplicates(subset=['Cdigo']).groupby('Mes')['Cdigo'].nunique()
print("Año de añalisis=",año)
Clientes_mes

año de añalisis= 2023


Mes
1    16
2    27
3     2
Name: Cdigo, dtype: int64

Para el caso de la variable Cdigo (Cliente), Se lleva a cabo un filtro por año y el conteo por mes de los cientes (a partir de función de agrupación (Groupby)), que se han obtenido en el tiempo desde la constitución de la empresa, hasta la fecha.

#### Variable = Vendedor

In [352]:
# Para determinar la cantidad de clientes por vendedor, lo determinaremos por año a partir del filtro por año
año_filtro.groupby('Vendedor')['Cdigo'].nunique()
clientes_vendedor = año_filtro.drop_duplicates(subset=['Cdigo']).groupby('Vendedor')['Cdigo'].nunique()
print("Año de añalisis=",año)
clientes_vendedor

Año de añalisis= 2023


Vendedor
1.0        2
2.0        5
3.0        5
4.0        4
5.0        1
8.0        4
9.0        3
11.0       1
13.0       1
14.0      12
16.0       2
17.0       1
18.0       2
19.0       1
9999.0     1
Name: Cdigo, dtype: int64

Para el caso de la variable Vendedor, como la Gerente requiere saber por cada vendedor cuantos clientes ha captado, entonces a partir de función de agrupación (Groupby) se visualiza por año los clientes por vendedor.

#### Analizar la distribución de las variables continuas

La dataset no cuenta con variables continuas por esta razón no aplica este item.

#### Evaluar la necesidad de usar Label encoding y/o  hot-encoding

Para la pregunta SMART que se plantea, se quiere identificar del total de clientes que han ingresado cuantos estan inactivos en la Empresa, es decir ya no volvieron a comprar. Para este caso se usara la variable "Bloqueado", la cual es una variable cualitativa nominal, en ese orden par el ejercicio de llevar esta variable a un modelo, se podria aplicar 

In [321]:
clientes_bloqueado= data.iloc[:,[0,10,14]].reset_index(drop=True)
clientes_bloqueado

Cdigo  Cupo de crdito Bloqueado
0       38643278               0        No
1      901242233        15000000        No
2      901242233        40500000        No
3      901242233        25000000        No
4      811039882        30000000        No
...          ...             ...       ...
2215  1129568780               0        No
2216    37392434               0        No
2217    80721672               0        No
2218    70288645               0        No
2219  1065808257               0        No

[2220 rows x 3 columns]

In [322]:
# se crea el objeto
onehotencoder = OneHotEncoder()
X = onehotencoder.fit_transform(clientes_bloqueado.Bloqueado.values.reshape(-1,1)).toarray()

X

array([[1., 0.],
       [1., 0.],
       [1., 0.],
       ...,
       [1., 0.],
       [1., 0.],
       [1., 0.]])

In [323]:
# Se agrega al dataset original
retirados = pd.DataFrame(X, columns = ["Bloqueado_"+str(int(i)) for i in range(X.shape[1])]) 
clientes_retirados = pd.concat([clientes_bloqueado, retirados], axis=1)
clientes_retirados

Cdigo  Cupo de crdito Bloqueado  Bloqueado_0  Bloqueado_1
0       38643278               0        No          1.0          0.0
1      901242233        15000000        No          1.0          0.0
2      901242233        40500000        No          1.0          0.0
3      901242233        25000000        No          1.0          0.0
4      811039882        30000000        No          1.0          0.0
...          ...             ...       ...          ...          ...
2215  1129568780               0        No          1.0          0.0
2216    37392434               0        No          1.0          0.0
2217    80721672               0        No          1.0          0.0
2218    70288645               0        No          1.0          0.0
2219  1065808257               0        No          1.0          0.0

[2220 rows x 5 columns]

In [324]:
# Calculamos Variance Inflation Factor(VIF).
# Cuantifica la  multicolinealidad en un análisis de regresión normal de mínimos cuadrados.

# Seleccionamos las variables que vamos a analizar 
X = clientes_retirados[['Cupo de crdito','Bloqueado_0', 'Bloqueado_1']]

# VIF dataframe
vif_data = pd.DataFrame()
vif_data["feature"] = X.columns

# calculating VIF for each feature

vif_data["VIF"] = [variance_inflation_factor(X.values, i)
                          for i in range(len(X.columns))]
print(vif_data)

          feature       VIF
0  Cupo de crdito  1.000363
1     Bloqueado_0  1.158120
2     Bloqueado_1  1.000023


In [325]:
#Calificación del VIF:
#VIF=1, Baja Multicollinearidad
#VIF<5, Moderada Multicollinearidad
#VIF>5, Multicollinearity Extrema (Se busca evitar)

#En este caso el VIF dio moderada multicollinearidad.